Move Edu program status history to FRAM

jeromehue opened the pull request:
### Description

Move EDU program status history to FRAM by using the `fram::RingArray` to store program status history entries. Additionally, improve the unit test for `RingArray` with the addition of a custom type `RingArray` test case.

Fixes #204

jeromehue committed:
WIP: Add fram-stored programStatusHistory

Add a new RingArray programStatusHistory variable,
and a new update() method for it.

jeromehue committed:
Add a test for a custom type RingArray

The only missing thing is to test that the fram is correctly written,
i.e something like `Require(fram::ram::memory[address] == value)`.

jeromehue committed:
Finish 'Fram is working' test for custom type RingArray

jeromehue committed:
Test the UpdateProgramStatusHistory() method

jeromehue committed:
Move EDU program status history to FRAM

jeromehue committed:
Fix cpp formatting

jeromehue committed:
Fix number of cached program status history entries

PatrickKa commented on Sts1CobcSw/Edu/ProgramStatusHistory.hpp:
```diff
@@ -2,18 +2,14 @@
 
-#include <strong_type/type.hpp>
-
 // clang-format off
```
Remove the clang-format comments. They are no longer needed

PatrickKa commented on Sts1CobcSw/Edu/ProgramStatusHistory.cpp:
```diff
@@ -1,24 +1,29 @@
 namespace sts1cobcsw::edu
 {
-RODOS::RingBuffer<ProgramStatusHistoryEntry, nProgramStatusHistoryEntries> programStatusHistory;
+constexpr auto programStatusHistorySection = framSections.Get<"eduProgramStatusHistory">();
```
Remove that variable. It is only used once.

PatrickKa commented on Sts1CobcSw/Edu/ProgramStatusHistory.cpp:
```diff
@@ -1,24 +1,29 @@
-        if(programStatusHistory.vals[i].startTime == startTime
-           and programStatusHistory.vals[i].programId == programId)
+        auto entry = programStatusHistory.Get(i);
+        if(entry.startTime == startTime and entry.programId == programId)
         {
-            programStatusHistory.vals[i].status = newStatus;
+            entry.status = newStatus;
+            programStatusHistory.Set(i, entry);
```
I just realized that this is not thread-safe in the sense that someone might call `PushBack()` between `Get()` and `Set()`. This would mean that the index `i` might point to a different location/entry. We somehow need to ensure that getting and setting the entry is atomic. The only way that I can think of right now that would achieve that is to add `RingArray<T>::FindAndReplace(predicate, newData/Element/Entry)` where `predicate` is a callable that takes a `T const &` and returns something convertible to `bool`.

PatrickKa commented on Sts1CobcSw/Edu/ProgramStatusHistory.hpp:
```diff
@@ -55,13 +51,14 @@ inline constexpr std::size_t serialSize<edu::ProgramStatusHistoryEntry> =
+                 nCachedProgramStatusHistoryEntries>
     programStatusHistory;
 
+inline constexpr auto nProgramStatusHistoryEntries = programStatusHistory.FramCapacity();
```
Remove this constant. It is never used and not much shorter than `programStatusHistory.FramCapacity()` itself.

PatrickKa reviewed (changes requested):
- Unit tests should be as self-contained as possible. Therefore, please remove the dependency on the EDU code and test `RingArray` with a custom type that is defined right in the source file.

jeromehue committed:
Apply some requested changes from review

Remove useless variables and clang-format comments.

jeromehue committed:
Fix unit test for RingArray

Remove dependency to Edu module.

PatrickKa committed:
Fix bug in `operator==(S, S)` and refactor FRAM ring array test

jeromehue committed:
Add a FindAndReplace method for RingArray

jeromehue committed:
Use FindAndReplace method in UpdateProgramStatusHistory

PatrickKa committed:
Refactor `RingArray`

- Shorten `FindAndReplace()` signature
- Rename `ComputeSize()` to `FramSize()`
- Add some comments

PatrickKa committed:
Add `RingArray::DoSize()`

PatrickKa committed:
Make `RingArray`'s static member variables `inline`

PatrickKa committed:
Fix `RingArray::Set()` not always writing to the cache

PatrickKa committed:
Fix `RingArray::PushBack` not always writing to the cache

Also fix the test since the cache entries remain even after resetting
the FRAM memory to zero.

PatrickKa committed:
Fix `RingArray::FindAndReplace()` calling public functions

Since `RingArray` uses the Thread-safe Interface pattern, public
functions must not call other public functions because that would lead
to recursive locking. Rodos' semaphores might support that but I think
it is cleaner and easier to reason about if we stick to the pattern and
ensure that we only ever enter the semaphore once.

`FindAndReplace()` used `Get()` and `Set()` so I had to make private
implementation functions (`DoGet()` and `DoSet()`) for them that do all
the work except for creating the `ScopeProtector`.

PatrickKa committed:
Fix warning and includes

There was a warning about a narrowing conversion from 64 to 32 bits, so
I changed the underlying type of `RingArray::RingIndex` from
`std::size_t` to that of `fram::Size` which is `std::uint32_t`.

PatrickKa committed:
Add `RingArray::IndexType` and `SizeType`

PatrickKa committed:
Ensure that `RingArray` caches the latest elements

PatrickKa committed:
Add tests for `RingArray::FindAndReplace()`

PatrickKa committed:
Fix some integer conversion warnings
